### PR TITLE
WIP: Win32 support for owned windows.

### LINKED ISSUE #3254: Window.Owner works differently than in WPF (Z order is not respected)
In WPF setting window `Owner` makes the child window stay on top of the parent. Quoting https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.owner?view=netframework-4.8:

```
To allow you to create a relationship between a child window and a parent window, Window supports the notion of ownership. Ownership is established when the Owner property of a window (the owned window) is set with a reference to another window (the owner window).

- Once this relationship is established, the following behaviors are exhibited:

- If an owner window is minimized, all its owned windows are minimized as well.

- If an owned window is minimized, its owner is not minimized.

- If an owner window is maximized, both the owner window and its owned windows are restored.

- An owner window can never cover an owned window.

- Owned windows that were not opened using ShowDialog are not modal. The user can still interact with the owner window.

- If you close an owner window, its owned windows are also closed.

- If an owned window was opened by its owner window using Show, and the owner window is closed, the owned window's Closing event is not raised.

```

In Avalonia `An owner window can never cover an owned window.` is not true and parented windows can be obscured by the parent window.

### PR DESCRIPTION
**Pull request by MarchingCube**
## What does the pull request do?

Add support for `Window.Owner` for Win32 platform. This should allow users to have behavior similar to WPF when running on Windows.

## What is the current behavior?

`Window.Owner` does nothing. Dialog windows do not show in taskbar.

## What is the updated/expected behavior with this PR?

On Win32 setting `Window.Owner` and then showing window will cause the newly shown window to stay on top of the parent. Closing parent will also close owned windows.

ControlCatalog has two new options in dialogs:
- open owned window
- open owned window that does not show in taskbar.

## How was the solution implemented (if it's not obvious)?

I had to cleanup style management in Win32 window implementation. Certain styles were set incorrectly (like `WS_EX_TOOLWINDOW` was always applied when toggling taskbar visbility) and some style setting was a bit convoluted.

If window cannot be resized I also disable maximize button (requested here: https://github.com/AvaloniaUI/Avalonia/issues/437).

Owner and `ShowDialog` are incompatible currently. WPF allows for using both, but almost always you would want to set both to have correct behavior. 
In Avalonia we have a few trickier scenarios:
- user sets `Owner` and then calls `ShowDialog` - `Owner` is ignored and new owner given to `ShowDialog` will be used. If user were to `Hide` that dialog and call `Show` - `Owner` will be used to set window owner. To make it more consistent we could set `Owner` when calling `ShowDialog`.
- user calls `ShowDialog` and then sets `Owner` - this currently throws as it is not really well defined what to do in that case (WPF does not support it either).

Missing part:
- I need similar solution like WPF to handle a case of a window that has no owner and user requests it to hide taskbar icon. This requires a new hidden window that we can use to set as temporary owner so it will disappear correctly from the taskbar. This is a bit niche feature so I didn't implement it right away.
- Other platforms are not implemented, but nothing will break and behavior will be the same as before.

## Checklist
- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
Modal dialogs now show in taskbar by default, but this more of a bugfix and not a break.

## Fixed issues
Fixes #3254

### CODE REVIEW

**Comment by Gillibald** [comment]
People often write `myDialog.ShowDialog(((IClassicDesktopStyleApplicationLifetime)applicationLifetime).MainWindow)` can me simplify this by changing the signature to `(...., Window = null)`?

**Comment by MarchingCube** [comment]
@Gillibald I guess it would be hard at the moment since win32 window implementation is not aware of the main window. And I also assume that @kekekeks does not want us hacking in checks if user application is using classic desktop lifetime and passing that around automatically.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 1]
```diff
@@ -333,7 +389,20 @@ public void SetTitle(string title)
 
         public virtual void Show()
         {
-            SetWindowLongPtr(_hwnd, (int)WindowLongParam.GWL_HWNDPARENT, IntPtr.Zero);
+            IntPtr ownerHandle = IntPtr.Zero;
+
+            // Window was potentially shown as dialog so we have to restore owner.
+            if (_ownerWindow != null)
```
Comment (Gillibald) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Is this a supported scenario to call `Show -> Close -> Show` multiple times? Once I close a window this is final in my understanding. For this use-case someone would call `Show -> Hide -> Activate`

Comment (MarchingCube) [reply src/Windows/Avalonia.Win32/WindowImpl.cs]: `Close` is final, but user can call `Show -> Hide -> Show` or even `Show -> Hide -> ShowDialog`. WPF seems to support this as well: https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.hide?view=netframework-4.8.

**Comment by Gillibald** [comment]
I think we should change the way we manage windows. We could change `IWindowingPlatform` to manage all window instances.
```
public interface IWindowingPlatform
{
    IWindowImpl CreateWindow();
    IEmbeddableWindowImpl CreateEmbeddableWindow();
    Window MainWindow { get; set; }
    IReadOnlyList<Window> Windows { get; }
}
```
That way application lifetime and window management are decoupled. A window always requires a windowing plaform. This change would allow us to remove the owner parameter from `ShowDialog`

**Comment by MarchingCube** [comment]
I think the problem stems from modal windows in a multi-window setup. Each of the main windows needs to be set as the owner of a show dialog. Otherwise shown dialogs will disable wrong window. And I think classic desktop lifetime does not require setting main window at the moment.

**Comment by Gillibald** [comment]
It's the same in WPF. If you call `ShowDialog` the `MainWindow` is used as the `Owner` by default but if you set `Owner` before `ShowDialog` `Owner` is used.

```
void ShowDialog()
{
    var owner = Owner;

    if(owner == null)
    {
        var windowingPlatform = AvaloniaLocator.Current.GetService<IWindowingPlatform>();
        owner = windowingPlatform.MainWindow;
    }
    ........
}
```

**Comment by MarchingCube** [comment]
I would need to check it but from what I understood from https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.owner?view=netframework-4.8 is that `Owner` and owner set by `ShowDialog` are two different concepts actually.

**Comment by Gillibald** [comment]
Hmm. With WPF the default behavior seems to be that all windows of the application become disabled until the dialog was closed. To change this you have to assign an owner. So my guess is that there is a hidden window that is used to manage all windows.

```
                     HiddenRoot
             /               |             \
 WindowA   MainWindow     DialogA
                              | 
                        DialogB
```
DialogB was opened with MeinWindow set as the owner and DialogA has no owner set.

**Comment by donandren** [comment]
not sure this is the right place, but shouldn't we use (for windows) set owner (GWL_HWNDPARENT style) also for popups by default (instead set WS_EX_TOPMOST to popup hwnd)?

**Comment by Gillibald** [comment]
@donandren That should make sure popups don't stay on the screen when the window is minimized / not visible with zero effort 👍

**Comment by MarchingCube** [comment]
@donandren Sounds good, although I wonder if that shouldn't be a separate PR. Since that popup change needs to be tested separately I would rather merge this change first and then do follow up work.

**Comment by MarchingCube** [comment]
@kekekeks Would you have a moment to review this PR?

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 2]
```diff
@@ -425,15 +493,6 @@ protected virtual IntPtr CreateWindowOverride(ushort atom)
                 IntPtr.Zero);
         }
 
-        bool ShouldIgnoreTouchEmulatedMessage()
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Again, `git blame` and anyone who would try to revert this commit in the future would be greatly confused.

**Review by kekekeks — CHANGES_REQUESTED** [review]
Our window style management is currently a mess. It's a bit more clean in X11 backend when we just recompute styles from scratch on property change. Not sure if that's in the scope of this PR, but you've removed one of the `UpdateWMStypes` calls...

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 3]
```diff
@@ -288,7 +314,37 @@ public void SetSystemDecorations(bool value)
                 return;
             }
 
-            UpdateWMStyles(()=> _decorated = value);
+            _decorated = value;
```
[the remark this answers is not in the document]
Comment (MarchingCube) [reply src/Windows/Avalonia.Win32/WindowImpl.cs]: I could convert it to a single unified place. When looking at this code I figured that both `UpdateWMStyles` and `ShowTaskbarIcon` are generally broken (setting pointless styles or even completely wrong styles) so I wanted to split it a  bit. Win32 being broken on its own as well and requiring hacks to apply certain properties (taskbar icons).

[the remark this answers is not in the document]
Comment (MarchingCube) [reply src/Windows/Avalonia.Win32/WindowImpl.cs]: And I think `UpdateWMStyles` was a hack for `SetSystemsDecorations` to begin with as it only cares about old decoration status and ignores `_resizable`.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 4]
```diff
@@ -161,14 +193,13 @@ public void Resize(Size value)
         {
             int requestedClientWidth = (int)(value.Width * Scaling);
             int requestedClientHeight = (int)(value.Height * Scaling);
-            UnmanagedMethods.RECT clientRect;
-            UnmanagedMethods.GetClientRect(_hwnd, out clientRect);
-           
+
+            UnmanagedMethods.GetClientRect(_hwnd, out var clientRect);
```
[the remark this answers is not in the document]
Comment (MarchingCube) [reply src/Windows/Avalonia.Win32/WindowImpl.cs]: I will remove formatting changes then.

**Comment by MarchingCube** [comment]
Will redo this on top of my reworked Win32 window code.

**Pull request closed without merging**